File: config/config-sbc2ha/src/main/java/com/dfi/sbc2ha/config/sbc2ha/definition/actuator/CoverConfig.java
package com.dfi.sbc2ha.config.sbc2ha.definition.actuator;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;

import java.time.Duration;

@Data
@JsonIgnoreProperties({"output", "momentaryTurnOn", "momentaryTurnOff"})
public class CoverConfig extends ActuatorConfig {
    /**
     * uniquely identifies this device in MQTT and Home Assistant.
     * required
     */
    String id;

    /**
     * Relay ID which opens a cover. It has to be set in output section.
     * required
     */
    @JsonProperty("open_relay")
    String openRelay;
    @JsonProperty("open_relay_bus_id")
    String openRelayBusId;
    @JsonProperty("open_relay_bus_type")
    String openRelayBusType;

    /**
     * Relay ID which closes a cover. It has to be set in output section.
     * Required
     */
    @JsonProperty("close_relay")
    String closeRelay;
    @JsonProperty("close_relay_bus_id")
    String closeRelayBusId;
    @JsonProperty("close_relay_bus_type")
    String closeRelayBusType;

    /**
     * Time to open a cover. Example 20 seconds.
     * Required
     */
    @JsonProperty("open_time")
    Duration openTime;

    /**
     * Time to close a cover. Example 30 sec.
     * Required
     */
    @JsonProperty("close_time")
    Duration closeTime;

    /**
     * Type of device to see in Home Assistant.
     * allowed_values: ['awning', 'blind', 'curtain', 'damper', 'door', 'garage', 'gate', 'shade', 'shutter', 'window']) -
     * Optional
     */
    @JsonProperty("device_class")
    String deviceClass;


}

File: config/editor/src/app/definition/actuator/CoverConfig.ts
import {ActuatorConfig} from "./ActuatorConfig";
import {Duration} from "../Duration";
import {OutputKindType} from "../enums/OutputKindType";
import {ActuatorType} from "../enums/ActuatorType";
import {Expose, Transform} from "class-transformer";
import {fromDuration, toDuration} from "../AppConfig";

export class CoverConfig extends ActuatorConfig {
    @Expose({name: "open_relay"})
    openRelay: string;

    @Expose({name: "open_relay_bus_id"})
    openRelayBusId: string;

    @Expose({name: "open_relay_bus_type"})
    openRelayBusType: string;

    @Expose({name: "close_relay"})
    closeRelay: string;

    @Expose({name: "close_relay_bus_id"})
    closeRelayBusId: string;

    @Expose({name: "close_relay_bus_type"})
    closeRelayBusType: string;

    @Transform(({value}) => toDuration(value), {toClassOnly: true})
    @Transform(({value}) => fromDuration(value), {toPlainOnly: true})
    @Expose({name: "open_time"})
    openTime: Duration;

    @Transform(({value}) => toDuration(value), {toClassOnly: true})
    @Transform(({value}) => fromDuration(value), {toPlainOnly: true})
    @Expose({name: "close_time"})
    closeTime: Duration;

    @Expose({name: "device_class"})
    deviceClass: string;


    constructor(name: string, outputType: ActuatorType, output: number, openRelay: string, openRelayBusId: string, openRelayBusType: string, closeRelay: string, closeRelayBusId: string, closeRelayBusType: string, openTime: Duration, closeTime: Duration, deviceClass: string) {
        super(OutputKindType.COVER, name, outputType, output);
        this.openRelay = openRelay;
        this.openRelayBusId = openRelayBusId;
        this.openRelayBusType = openRelayBusType;
        this.closeRelay = closeRelay;
        this.closeRelayBusId = closeRelayBusId;
        this.closeRelayBusType = closeRelayBusType;
        this.openTime = openTime;
        this.closeTime = closeTime;
        this.deviceClass = deviceClass;
    }
}
